### PR TITLE
Use vendored spans in tags generation

### PR DESCRIPTION
**Pull request by kevingosse**
## Summary of changes

In our tags code, use spans instead of arrays on .net framework and netstandard.

## Reason for change

It won't actually speed up anything (spans are used here on netcoreapp to benefit from a compiler optimization that, afaik, does not exist on netfx), but this is an experiment to validate that we can seamlessly use vendored spans on netfx and real spans on netcore. Plus, it makes this code much simpler.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5298)](https://github.com/DataDog/dd-trace-dotnet/tree/e898ed121562ed0a410dda26b83ec3255b869f26)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/a8cb7e356d1aab3f3f4d0f170c00df0d5aaf7b9e)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5298) - mean (74ms)  : 66, 83
     .   : milestone, 74,
    master - mean (74ms)  : 65, 83
     .   : milestone, 74,

    section CallTarget+Inlining+NGEN
    This PR (5298) - mean (999ms)  : 969, 1029
     .   : milestone, 999,
    master - mean (985ms)  : 961, 1008
     .   : milestone, 985,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5298) - mean (113ms)  : 109, 117
     .   : milestone, 113,
    master - mean (118ms)  : 96, 140
     .   : milestone, 118,

    section CallTarget+Inlining+NGEN
    This PR (5298) - mean (725ms)  : 688, 762
     .   : milestone, 725,
    master - mean (713ms)  : 691, 735
     .   : milestone, 713,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5298) - mean (95ms)  : 91, 99
     .   : milestone, 95,
    master - mean (95ms)  : 91, 98
     .   : milestone, 95,

    section CallTarget+Inlining+NGEN
    This PR (5298) - mean (677ms)  : 652, 702
     .   : milestone, 677,
    master - mean (671ms)  : 647, 694
     .   : milestone, 671,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5298) - mean (188ms)  : 185, 192
     .   : milestone, 188,
    master - mean (188ms)  : 182, 193
     .   : milestone, 188,

    section CallTarget+Inlining+NGEN
    This PR (5298) - mean (1,070ms)  : 1045, 1095
     .   : milestone, 1070,
    master - mean (1,063ms)  : 1038, 1089
     .   : milestone, 1063,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5298) - mean (272ms)  : 268, 276
     .   : milestone, 272,
    master - mean (269ms)  : 262, 276
     .   : milestone, 269,

    section CallTarget+Inlining+NGEN
    This PR (5298) - mean (867ms)  : 846, 889
     .   : milestone, 867,
    master - mean (861ms)  : 839, 884
     .   : milestone, 861,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5298) - mean (260ms)  : 256, 263
     .   : milestone, 260,
    master - mean (258ms)  : 250, 265
     .   : milestone, 258,

    section CallTarget+Inlining+NGEN
    This PR (5298) - mean (852ms)  : 833, 871
     .   : milestone, 852,
    master - mean (855ms)  : 834, 876
     .   : milestone, 855,

```

File: tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs [hunk 1]
```diff
@@ -430,11 +430,7 @@ private void WriteTag(ref byte[] bytes, ref int offset, string key, string value
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-#if !NETCOREAPP
```
Comment (daniel-romano-DD) [inline tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs:433]: Only to get rid of these #ifs it's already worth it

File: tracer/src/Datadog.Trace/GlobalUsings.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// <copyright file="GlobalUsings.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+global using ThrowHelper = Datadog.Trace.Util.ThrowHelper;
+
+#if NETFRAMEWORK || NETSTANDARD2_0
```
Comment (daniel-romano-DD) [inline tracer/src/Datadog.Trace/GlobalUsings.cs:8]: Nice

**Review by tonyredondo — APPROVED** [review]
LGTM

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5298)](https://github.com/DataDog/dd-trace-dotnet/tree/f7c26c2986db214cd9567550c1593bb36e74c5b4)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/82498f002fcbc6dd9bacca42a44850c72cea10a5)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5298) (11.129M)   : 0, 11128579
    master (11.387M)   : 0, 11386870
    benchmarks/2.9.0 (11.099M)   : 0, 11099186

    section Automatic
    This PR (5298) (7.636M)   : 0, 7636217
    master (7.766M)   : 0, 7766388
    benchmarks/2.9.0 (7.823M)   : 0, 7823285

    section Trace stats
    This PR (5298) (7.808M)   : 0, 7808046
    master (8.071M)   : 0, 8071440

    section Manual
    This PR (5298) (9.568M)   : 0, 9568459
    master (9.806M)   : 0, 9806449

    section Manual + Automatic
    This PR (5298) (7.196M)   : 0, 7195572
    master (7.297M)   : 0, 7297336

    section Version Conflict
    This PR (5298) (6.382M)   : 0, 6382456
    master (6.603M)   : 0, 6602958

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5298) (9.414M)   : 0, 9413611
    master (9.706M)   : 0, 9706267
    benchmarks/2.9.0 (9.709M)   : 0, 9708719

    section Automatic
    This PR (5298) (6.663M)   : 0, 6662548
    master (6.592M)   : 0, 6591996

    section Trace stats
    This PR (5298) (6.934M)   : 0, 6934435
    master (6.912M)   : 0, 6912452

    section Manual
    This PR (5298) (8.166M)   : 0, 8165580
    master (8.134M)   : 0, 8134221

    section Manual + Automatic
    This PR (5298) (6.300M)   : 0, 6299560
    master (6.177M)   : 0, 6177016

    section Version Conflict
    This PR (5298) (5.571M)   : 0, 5570770
    master (5.622M)   : 0, 5621557

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5298) (9.824M)   : 0, 9823797
    master (10.038M)   : 0, 10037686
    benchmarks/2.9.0 (10.184M)   : 0, 10183565

    section Automatic
    This PR (5298) (7.094M)   : 0, 7093793
    master (7.146M)   : 0, 7146271
    benchmarks/2.9.0 (7.254M)   : 0, 7253763

    section Trace stats
    This PR (5298) (7.329M)   : 0, 7329116
    master (7.389M)   : 0, 7389237

    section Manual
    This PR (5298) (8.735M)   : 0, 8735399
    master (8.795M)   : 0, 8795107

    section Manual + Automatic
    This PR (5298) (6.785M)   : 0, 6784700
    master (6.908M)   : 0, 6907539

    section Version Conflict
    This PR (5298) (6.162M)   : 0, 6161763
    master (6.366M)   : 0, 6365847

```

**Pull request merged**